Keeps environmentMapReady promise in pending state after reset.

amarchenko opened the pull request:
#### 06fa32413283ac063828168d6fc5ec804c0661e3
<pre>
Keeps environmentMapReady promise in pending state after reset.
<a href="https://bugs.webkit.org/show_bug.cgi?id=284753">https://bugs.webkit.org/show_bug.cgi?id=284753</a>
<a href="https://rdar.apple.com/141467143">rdar://141467143</a>

Reviewed by Ada Chan.

environmentMapReadyPromise must be in pending state after resetting
environmentMap with an empty string. That behavior matches readyPromise
that signals models source loading.
In this particular fix we check whether a URL for environmentMap is
provided or not before resolving the promise.

There are no changes in LayoutTests/model-element/model-element-environment-map-ready.html
since there is no reliable way to verify that a Promise
is in pending state. The only thing we can do is to make sure
that modelElement.environmentMapReady promise has changed after URL reset.
That verification happens in `&lt;model&gt; rejects the environmentMapReady promise when its environmentMap load is aborted.`

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::didFinishEnvironmentMapLoading):

Canonical link: <a href="https://commits.webkit.org/288124@main">https://commits.webkit.org/288124@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8808e06ddc60968d08ad5af62df7dc0fa2c4e6d6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81701 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1227 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35655 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86245 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32696 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83807 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1260 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9047 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63748 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21468 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84771 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/913 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74353 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44034 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/812 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28527 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31150 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72200 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29129 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87684 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8941 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6349 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72079 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9126 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70172 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71309 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17815 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15394 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14309 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8892 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14424 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8733 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12256 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10541 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 reviewed (approved):
Do we already have an existing test for resetting the environment map?

amarchenko committed:
Keeps environmentMapReady promise in pending state after reset.
https://bugs.webkit.org/show_bug.cgi?id=284753
rdar://141467143

Reviewed by Ada Chan.

environmentMapReadyPromise must be in pending state after resetting
environmentMap with an empty string. That behavior matches readyPromise
that signals models source loading.
In this particular fix we check whether a URL for environmentMap is
provided or not before resolving the promise.

There are no changes in LayoutTests/model-element/model-element-environment-map-ready.html
since there is no reliable way to verify that a Promise
is in pending state. The only thing we can do is to make sure
that modelElement.environmentMapReady promise has changed after URL reset.
That verification happens in `<model> rejects the environmentMapReady promise when its environmentMap load is aborted.`

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::didFinishEnvironmentMapLoading):

Canonical link: https://commits.webkit.org/288124@main
